File: gems/canvas_quiz_statistics/lib/canvas_quiz_statistics/analyzers/multiple_dropdowns.rb
#
# Copyright (C) 2014 Instructure, Inc.
#
# This file is part of Canvas.
#
# Canvas is free software: you can redistribute it and/or modify it under
# the terms of the GNU Affero General Public License as published by the Free
# Software Foundation, version 3 of the License.
#
# Canvas is distributed in the hope that it will be useful, but WITHOUT ANY
# WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
# A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
# details.
#
# You should have received a copy of the GNU Affero General Public License along
# with this program. If not, see <http://www.gnu.org/licenses/>.
#
module CanvasQuizStatistics::Analyzers
  class MultipleDropdowns < FillInMultipleBlanks
    inherit_metrics :fill_in_multiple_blanks_question

    private

    def analyze_response_for_blank(response, blank, answer_set)
      answer_id = response.fetch(answer_key(blank), nil).to_s

      answer = if answer_id.present?
        answer_set[:answers].detect { |a| "#{a[:id]}" == answer_id }
      end

      answer ||= generate_incorrect_answer({
        id: MissingAnswerKey,
        text: MissingAnswerText,
        in: answer_set
      })

      answer[:responses] += 1 if answer.present?
    end
  end
end

File: client_apps/canvas_quizzes/apps/events/js/views/question_inspector/answers/multiple_dropdowns.js
/** @jsx React.DOM */
/*
 * Copyright (C) 2014 - present Instructure, Inc.
 *
 * This file is part of Canvas.
 *
 * Canvas is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, version 3 of the License.
 *
 * Canvas is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Affero General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */

define(function(require) {
  var React = require('old_version_of_react_used_by_canvas_quizzes_client_apps')
  var K = require('../../../constants')
  var NO_ANSWER = require('jsx!./no_answer')
  var keys = Object.keys

  var MultipleDropdowns = React.createClass({
    statics: {
      questionTypes: [K.Q_MULTIPLE_DROPDOWNS]
    },

    render: function() {
      var answer = this.props.answer
      var question = this.props.question
      var answers = this.props.question.answers

      return (
        <table>
          {keys(answer).map(function(blank) {
            var answerText =
              answers.filter(function(originalAnswer) {
                return '' + originalAnswer.id === answer[blank]
              })[0] || {}

            return (
              <tr key={'blank' + blank}>
                <th scope="row">{blank}</th>
                <td>{answerText.text || NO_ANSWER}</td>
              </tr>
            )
          })}
        </table>
      )
    }
  })

  return MultipleDropdowns
})
